### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by jeanpaulvanr**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/data/trello_items.py [hunk 1]
```diff
@@ -29,19 +26,19 @@ def get_items(status):
                                                                   
     return list_of_cards
 
-def get_open_items():
+def get_do_items():
```
Comment (CorndelWithSoftwireDevOps) [inline todo_app/data/trello_items.py:29]: The main idea was to have the logic to split the items into Do/Doing/Done in the ViewModel, not here, but this works as well

File: todo_app/data/test_view_model.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from todo_app.data.item import Item
+from todo_app.data.view_model import ViewModel 
+import pytest
+
+# Arrange
+@pytest.fixture
+def do_items_list(): 
+    return [
+        Item(1, "My Todo Item 1", "To Do"),
```
Comment (CorndelWithSoftwireDevOps) [inline todo_app/data/test_view_model.py:9]: I don't see the benefit of adding 3 items in each list, your tests work just as well with only one, right?

File: README.md [hunk 3]
```diff
@@ -51,6 +51,12 @@ The project uses a virtual environment to isolate package dependencies. To creat
 $ poetry install
 ```
 
+You'll then need to run 'poetry add requests' (as it's not currently tracked by Poetry):
+
+```bash
+$ poetry add requests  # (first time only)
```
[the remark this answers is not in the document]
Comment (jeanpaulvanr) [reply README.md:57]: Thanks to module 5 guidance, I understand what you mean now. Thanks. Removed it now.

File: README.md [hunk 4]
```diff
@@ -92,3 +98,143 @@ You should see output similar to the following:
...
+
+
+## Testing the App
+
+### How to Set Up for Testing
+
+You need to add pytest as a dependency of our project.
+
+```bash
+$ poetry add pytest --dev
```
[the remark this answers is not in the document]
Comment (jeanpaulvanr) [reply README.md:110]: Thanks to module 5 guidance, I understand what you mean now. Thanks. Removed it now.

File: README.md [hunk 5]
```diff
@@ -92,3 +98,143 @@ You should see output similar to the following:
...
+This should download it and also update pyproject.toml for you.
+
+You can now run either the Unit Tests or the Integration Tests or both. Please see below.
+
+### Unit Testing
+
+To run the Unit Tests on their own:
+
+```bash
+$ poetry run pytest todo_app\data\test_view_model.py
```
[the remark this answers is not in the document]
Comment (jeanpaulvanr) [reply README.md:121]: Okay. Thanks. I've changed this.

File: todoapp-ansible-playbook.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
+-- name: Install dependencies
+  hosts: webservers
+  remote_user: ec2-user
+  vars_prompt:
+
+  - name: trello_board_id
+    prompt: What is your Trello Board ID?
+    private: no
```
[the remark this answers is not in the document]
Comment (jeanpaulvanr) [reply todoapp-ansible-playbook.yml:8]: Thanks. I've changed this.

File: todoapp-ansible-playbook.yml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    ansible.builtin.git:
+      repo: https://github.com/jeanpaulvanr/Devops-Course-Starter.git
+      dest: /opt/todoapp
+      single_branch: yes
+      version: module_4
+      force: yes
+
+  - name: Change the working directory to opt/todoapp before executing the poetry commands
+    ansible.builtin.shell:
+      cmd: /home/ec2-user/.local/bin/poetry add requests
```
[the remark this answers is not in the document]
Comment (jeanpaulvanr) [reply todoapp-ansible-playbook.yml:54]: Thanks. I understand this much better now. I've changed this.

File: todo_app/data/trello_items.py (hunk 1 above)
Comment (jeanpaulvanr) [reply todo_app/data/trello_items.py]: Thanks. Jack has had me reset my code as he could foresee problems with modules further down the line so this code is no longer applicable btw.

File: todo_app/data/test_view_model.py (hunk 2 above)
Comment (jeanpaulvanr) [reply todo_app/data/test_view_model.py:9]: Fair point. I've reduced it to 1. Thanks.

**Pull request merged**